### PR TITLE
Read SPK files directly

### LINKED ISSUE #15: Support SPK Type 13 - Hermite unequal Time Steps
[Documentation](https://naif.jpl.nasa.gov/pub/naif/toolkit_docs/C/req/spk.html#Type%2013:%20Hermite%20Interpolation%20---%20Unequal%20Time%20Steps)

_Note:_ this is a set of discrete states which can be used, it seems, to _build_ a Hermite interpolation scheme.

# Purpose
Be able to convert a BSP file of Type 13 to an ANISE trajectory file.

# SPICE organization
```
                                       +--------+
                                       |  x(1)  |
                                   /   +--------+
                                  /    |  y(1)  |
                                 /     +--------+
                                /      |  z(1)  |
   +-----------------------+   /       +--------+
   | State 1               |  <        |dx(1)/dt|
   +-----------------------+   \       +--------+
   | State 2               |    \      |dy(1)/dt|
   +-----------------------+     \     +--------+
               .                  \    |dz(1)/dt|
               .                       +--------+
               .
   +-----------------------+
   | State N               |
   +-----------------------+
   | Epoch 1               |
   +-----------------------+
   | Epoch 2               |
   +-----------------------+
               .
               .
               .
   +-----------------------+
   | Epoch N               |
   +-----------------------+
   | Epoch 100             | (First directory)
   +-----------------------+
               .
               .
               .
   +-----------------------+
   | Epoch ((N-1)/100)*100 | (Last directory)
   +-----------------------+
   | Window size - 1       |
   +-----------------------+
   | Number of states      |
   +-----------------------+
```

# Test requirements
- [x] ~Find a public relevant BSP (Cassini?)~ Build a BSP from GMAT
- [x] Ensure correct parsing
- [x] Ensure correct copying of the data
- [x] Ensure correct computation of states

### LINKED ISSUE #30: SPICE BSP files should be usable without needing the ANISE file equivalent
# Use case
Although the ANISE file format provides a number of advantages compared to the BSP files, requiring users to convert their data into ANISE before using this library is a large overhead.

# Effort
ANISE.rs already supports reading BSP files of some types. However, that implementation requires memory allocations. These should not be required.

Hence, the crux of this issue is to rewrite the SPK parser to not perform any memory allocations. This issue also includes providing a rewrite of some of the SPK related functions (exact list to be determined from [this list](https://naif.jpl.nasa.gov/pub/naif/toolkit_docs/C/cspice/index.html#S)). These functions must be behind a specific crate feature, e.g. `spice-interop`.

# Verification

_TDB_: from the list of functions above, the tests should require `rust-spice` by @GregoireHENRY and make sure that the SPK functions selects return the same values. This will likely require #26 too.

### PR DESCRIPTION
**Pull request by ChristopherRabotin**
This replaces the ASN1 based Ephemeris and Orientation data. There may be a future use case for orientation data that is in ASN.1 encoding, but SPK files are way too common in the industry to require folks to switch from one file type to another.

This structure whoever removes the need to `unload` (or `unfurnsh`) files thanks to the proper use of lifetimes in Rust.

I will need to check whether one can deliberately violate these lifetimes in C (they might!).

TODO:
- [x] Better tests using the parquet file
- [x] Setup CSPICE in the pipeline using the shell script
- [x] Fix Github test
- [x] Add CLI testing

Closes #15 . Closes #30

### CODE REVIEW

**Comment by ChristopherRabotin** [comment]
Works, when using _exactly_ the same algorithm as SPICE.

Validation of a GMAT-generated SPK file, sampled 999 times over 3h 20 mins:

```
shape: (1, 7)
┌─────────────────────┬─────────────────────┬──────────────────────┬────────────────────────┬─────────────────────┬─────────────────────┬─────────────────────┐
│ min rel err (km) OK ┆ q25 rel err (km) OK ┆ mean rel err (km) OK ┆ median rel err (km) OK ┆ q75 rel err (km) OK ┆ q99 rel err (km) OK ┆ max rel err (km) OK │
│ ---                 ┆ ---                 ┆ ---                  ┆ ---                    ┆ ---                 ┆ ---                 ┆ ---                 │
│ f64                 ┆ f64                 ┆ f64                  ┆ f64                    ┆ f64                 ┆ f64                 ┆ f64                 │
╞═════════════════════╪═════════════════════╪══════════════════════╪════════════════════════╪═════════════════════╪═════════════════════╪═════════════════════╡
│ 0.0                 ┆ 6.2677e-16          ┆ 3.7227e-13           ┆ 1.6674e-15             ┆ 6.0504e-13          ┆ 1.5941e-12          ┆ 1.6198e-12          │
└─────────────────────┴─────────────────────┴──────────────────────┴────────────────────────┴─────────────────────┴─────────────────────┴─────────────────────┘
shape: (1, 7)
┌──────────────────┬──────────────────┬───────────────────┬─────────────────────┬──────────────────┬──────────────────┬──────────────────┐
│ min abs err (km) ┆ q25 abs err (km) ┆ mean abs err (km) ┆ median abs err (km) ┆ q75 abs err (km) ┆ q99 abs err (km) ┆ max abs err (km) │
│ ---              ┆ ---              ┆ ---               ┆ ---                 ┆ ---              ┆ ---              ┆ ---              │
│ f64              ┆ f64              ┆ f64               ┆ f64                 ┆ f64              ┆ f64              ┆ f64              │
╞══════════════════╪══════════════════╪═══════════════════╪═════════════════════╪══════════════════╪══════════════════╪══════════════════╡
│ 0.0              ┆ 6.6034e-9        ┆ 2.3705e-7         ┆ 3.2869e-8           ┆ 2.0364e-7        ┆ 0.000002         ┆ 0.000003         │
└──────────────────┴──────────────────┴───────────────────┴─────────────────────┴──────────────────┴──────────────────┴──────────────────┘
shape: (59, 6)
┌─────────────────────┬─────────────────────┬──────────────────┬──────────────────────┬───────────────────┬───────────┐
│ absolute error (km) ┆ relative error (km) ┆ File delta T (s) ┆ source frame         ┆ destination frame ┆ component │
│ ---                 ┆ ---                 ┆ ---              ┆ ---                  ┆ ---               ┆ ---       │
│ f64                 ┆ f64                 ┆ f64              ┆ str                  ┆ str               ┆ str       │
╞═════════════════════╪═════════════════════╪══════════════════╪══════════════════════╪═══════════════════╪═══════════╡
│ 0.000002            ┆ 1.5348e-12          ┆ 3252.000002      ┆ body -10000001 J2000 ┆ Earth J2000       ┆ Y         │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┤
│ 0.000002            ┆ 1.4980e-12          ┆ 3348.000002      ┆ body -10000001 J2000 ┆ Earth J2000       ┆ Y         │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┤
│ 0.000003            ┆ 1.4459e-12          ┆ 3444.000002      ┆ body -10000001 J2000 ┆ Earth J2000       ┆ Y         │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┤
│ 0.000002            ┆ 1.4383e-12          ┆ 3456.000002      ┆ body -10000001 J2000 ┆ Earth J2000       ┆ Y         │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┤
│ ...                 ┆ ...                 ┆ ...              ┆ ...                  ┆ ...               ┆ ...       │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┤
│ 0.000002            ┆ 9.7376e-13          ┆ 11244.000005     ┆ body -10000001 J2000 ┆ Earth J2000       ┆ Y         │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┤
│ 0.000002            ┆ 1.0935e-12          ┆ 11340.000005     ┆ body -10000001 J2000 ┆ Earth J2000       ┆ Y         │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┤
│ 0.000002            ┆ 1.2020e-12          ┆ 11436.000006     ┆ body -10000001 J2000 ┆ Earth J2000       ┆ Y         │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┤
│ 0.000002            ┆ 1.2981e-12          ┆ 11532.000006     ┆ body -10000001 J2000 ┆ Earth J2000       ┆ Y         │
└─────────────────────┴─────────────────────┴──────────────────┴──────────────────────┴───────────────────┴───────────┘

```

File: src/context/mod.rs [hunk 1]
```diff
@@ -8,11 +8,138 @@
...
-pub mod save;
+use hifitime::Epoch;
+
+use crate::errors::AniseError;
+use crate::naif::spk::summary::SPKSummaryRecord;
+use crate::naif::{BPC, SPK};
+use core::fmt;
+use log::error;
+
+pub const MAX_LOADED_FILES: usize = 32;
```
Comment (ChristopherRabotin) [inline src/context/mod.rs:19]: Consider making this a compile-time flag via build.rs.

File: src/math/polyfit/cdemo/Makefile [hunk 2]
```diff
@@ -0,0 +1,4 @@
+it:
```
Comment (ChristopherRabotin) [inline src/math/polyfit/cdemo/Makefile:1]: This whole module can probably be removed.

File: src/math/polyfit/mod.rs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+ *
+ * Documentation: https://nyxspace.com/
+ */
+
+use core::ops::{Index, IndexMut};
+
+/// A FixedArray is a a way around allocating vectors when we don't know the exact size at compile time.
+/// This will be made obsolete when https://github.com/rust-lang/rust/issues/60551 is merged into rust stable.
+#[derive(Copy, Clone, Debug)]
+pub struct F64TupleArray<const N: usize, const S: usize>([[f64; N]; S]);
```
Comment (ChristopherRabotin) [inline src/math/polyfit/mod.rs:16]: Keeping this for posteriori to enable building a Hermite interpolation scheme without the intermediate array that contains the ordinates and first derivatives (abscissas / epochs are contiguous so we don't store them in an array).

A struct of an underlying `&'a [f64]` can be used to skip items such that the indexing is every sixth element for example.

**Review by pwnorbitals — APPROVED** [review]

**Pull request merged**